com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/AndroidBundler.java

```diff
@@ -12,23 +12,670 @@
+            log("Copy debug symbols " + symbolExe);
+            BundleHelper.throwIfCanceled(canceled);
+            FileUtils.copyFile(exe, symbolExe);
+        }
+
+        BundleHelper.throwIfCanceled(canceled);
+
+        File proguardMapping = new File(FilenameUtils.concat(extenderExeDir, FilenameUtils.concat(architectures.get(0).getExtenderPair(), "mapping.txt")));
+        if (proguardMapping.exists()) {
+            File symbolMapping = new File(symbolsDir, proguardMapping.getName());
+            FileUtils.copyFile(proguardMapping, symbolMapping);
+        }
+    }
+
+    /**
+    * Cleanup bundle folder from intermediate folders and artifacts.
+    */
+    private static void cleanupBundleFolder(Project project, File outDir, File androidResDir, ICanceled canceled) throws IOException, CompileExceptionError {
+        log("Cleanup bundle folder");
+        BundleHelper.throwIfCanceled(canceled);
+        FileUtils.deleteDirectory(androidResDir);
+        FileUtils.deleteDirectory(new File(outDir, "aab"));
+    }
+
+    private static File createAAB(Project project, File outDir, BundleHelper helper, ICanceled canceled) throws IOException, CompileExceptionError {
+        BundleHelper.throwIfCanceled(canceled);
+
+        // STEP 1. copy android resources (icons, extension resources and manifest)
+        File manifestFile = helper.copyOrWriteManifestFile(Platform.Armv7Android, outDir);
+        File androidResDir = copyResources(project, outDir, helper, canceled);
+
+        // STEP 2. Use aapt2 to compile resources (to *.flat files)
+        File compiledResDir = aapt2CompileResources(project, androidResDir, canceled);
+
+        // STEP 3. Use aapt2 to create an APK containing resource files in protobuf format
+        File apk = aapt2LinkResources(project, outDir, compiledResDir, manifestFile, canceled);
+
+        // STEP 4. Extract protobuf files from the APK and create base.zip (manifest, assets, dex, res, lib, *.pb etc)
+        File baseZip = createAppBundleBaseZip(project, outDir, apk, canceled);
+
+        // STEP 5. Use bundletool to create AAB from base.zip
+        File baseAab = createBundle(project, outDir, baseZip, canceled);
+
+        //STEP 6. Sign AAB file
+        signFile(project, baseAab, canceled);
+
+        // STEP 7. Copy debug symbols
+        copySymbols(project, outDir, canceled);
+
+        // STEP 8. Cleanup bundle folder from intermediate folders and artifacts.
+        cleanupBundleFolder(project, outDir, androidResDir, canceled);
+
+        return baseAab;
+    }
+
+
+
+    /**
+     * Create a universal APK set. A universal APK includes all of the app's
+     * code and resources such that the APK is compatible with all device
+     * configurations the app supports.
+     */
+    private static File createUniversalApks(Project project, File aab, File outDir, ICanceled canceled) throws IOException, CompileExceptionError {
```

Comment:
This is the first step when creating an APK from an AAB file. An APK set contains one or more APKs. In this case we generate a single universal APK in the APK set.